### PR TITLE
Make warning a class rather than an interface

### PR DESCRIPTION
**Pull request by rictic**
- [x] CHANGELOG.md has been updated

One last thing question here. `Warning` is a bit of an awkward name for this, as warnings can have error severity or info severity, etc. If there's a better name this would be the time to change it, as we'll need to do one last pass to update all uses anyways. The Microsoft world calls them Diagnostics, which I find a touch too abstract. The atom linter calls them Messages, which I kinda like.

### CODE REVIEW

**Comment by justinfagnani** [comment]
I think Warning is an ok name, though I see your point. For our use, even errors are situations we _think_ are errors, but due to the über-dynamic nature of the platform, may not be. We can think of a Warning with severity Error as a "Warning: This looks like an Error".

File: src/model/warning.ts [hunk 1]
```diff
@@ -12,12 +12,114 @@
...
+    return {
+      code: this.code,
+      message: this.message,
+      severity: this.severity,
+      sourceRange: this.sourceRange,
+    };
+  }
 }
 
 export enum Severity {
```
Comment (justinfagnani) [inline src/model/warning.ts]: We really need some definition of the levels.

File: src/model/warning.ts [hunk 2]
```diff
@@ -12,12 +12,114 @@
...
+    const colorize = opts.color ? this._severityToColorFunction(this.severity) :
+                                  (s: string) => s;
+    const severity = this._severityToString(colorize);
+
+    if (!this.sourceRange || !this._parsedDocument) {
+      return `INTERNAL ERROR: Tried to print a '${this.code}' ` +
+          `warning without a source range and/or parsed document. ` +
+          `Please report this!\n` +
+          `     https://github.com/Polymer/polymer-analyzer/issues/new\n` +
+          `${this._severityToString(colorize)} ` +
```
Comment (justinfagnani) [inline src/model/warning.ts]: use `severity`

File: src/model/warning.ts [hunk 3]
```diff
@@ -12,12 +12,114 @@
...
+  }
+
+  toString(options: Partial<WarningStringifyOptions> = {}): string {
+    const opts:
+        WarningStringifyOptions = {...defaultPrinterOptions, ...options};
+    const colorize = opts.color ? this._severityToColorFunction(this.severity) :
+                                  (s: string) => s;
+    const severity = this._severityToString(colorize);
+
+    if (!this.sourceRange || !this._parsedDocument) {
```
Comment (justinfagnani) [inline src/model/warning.ts]: why can't you do this check on construction? It'll be closer to the real error and a stack might be useful.

File: src/model/warning.ts (hunk 1 above)
Comment (rictic) [inline src/model/warning.ts]: Agreed, later.

File: src/model/warning.ts (hunk 2 above)
Comment (rictic) [inline src/model/warning.ts]: removed

File: src/model/warning.ts (hunk 3 above)
Comment (rictic) [inline src/model/warning.ts]: +1, done

**Comment by rictic** [comment]
Got a verbal approval from Justin on this. Pinkie swear.

**Pull request merged**